package eu.kanade.tachiyomi.ui.browse.source.browse

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Save
import androidx.compose.material3.Button
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LocalTextStyle
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.runtime.withFrameNanos
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.unit.dp
import eu.kanade.presentation.components.AdaptiveSheet
import eu.kanade.presentation.util.isTvUi
import eu.kanade.tachiyomi.source.model.FilterList
import kotlinx.collections.immutable.ImmutableList
import kotlinx.collections.immutable.toImmutableList
import tachiyomi.core.common.preference.TriState
import tachiyomi.domain.source.model.EXHSavedSearch
import tachiyomi.i18n.MR
import tachiyomi.presentation.core.components.CheckboxItem
import tachiyomi.presentation.core.components.CollapsibleBox
import tachiyomi.presentation.core.components.HeadingItem
import tachiyomi.presentation.core.components.Scroller.STICKY_HEADER_KEY_PREFIX
import tachiyomi.presentation.core.components.SelectItem
import tachiyomi.presentation.core.components.SortItem
import tachiyomi.presentation.core.components.TextItem
import tachiyomi.presentation.core.components.TriStateItem
import tachiyomi.presentation.core.i18n.stringResource
import eu.kanade.tachiyomi.animesource.model.AnimeFilter as Filter

@Composable
fun SourceFilterDialog(
    onDismissRequest: () -> Unit,
    filters: FilterList,
    onReset: () -> Unit,
    onFilter: () -> Unit,
    onUpdate: (FilterList) -> Unit,
    // SY -->
    startExpanded: Boolean,
    savedSearches: ImmutableList<EXHSavedSearch>,
    onSave: () -> Unit,
    onSavedSearch: (EXHSavedSearch) -> Unit,
    onSavedSearchPress: (EXHSavedSearch) -> Unit,
    // SY <--
    // KMK -->
    onSavedSearchPressDesc: String,
    shouldShowSavingButton: Boolean = true,
    // KMK <--
) {
    val updateFilters = { onUpdate(filters) }
    val resetFocusRequester = remember { FocusRequester() }
    val isTvUi = isTvUi()

    LaunchedEffect(isTvUi) {
        if (isTvUi) {
            withFrameNanos { }
            resetFocusRequester.requestFocus()
        }
    }

    AdaptiveSheet(
        onDismissRequest = onDismissRequest,
        content = {
            LazyColumn {
                stickyHeader(
                    key = "$STICKY_HEADER_KEY_PREFIX-title",
                ) {
                    Row(
                        modifier = Modifier
                            .background(MaterialTheme.colorScheme.background)
                            .padding(8.dp),
                    ) {
                        TextButton(
                            modifier = Modifier.focusRequester(resetFocusRequester),
                            onClick = onReset,
                        ) {
                            Text(
                                text = stringResource(MR.strings.action_reset),
                                style = LocalTextStyle.current.copy(
                                    color = MaterialTheme.colorScheme.primary,
                                ),
                            )
                        }

                        Spacer(modifier = Modifier.weight(1f))

                        // KMK -->
                        if (shouldShowSavingButton) {
                            // KMK <--
                            // SY -->
                            IconButton(
                                onClick = onSave,
                            ) {
                                Icon(
                                    imageVector = Icons.Default.Save,
                                    contentDescription = stringResource(MR.strings.action_save),
                                    tint = MaterialTheme.colorScheme.onBackground,
                                )
                            }
                            // SY <--
                        }

                        Button(
                            onClick = {
                                onFilter()
                                onDismissRequest()
                            },
                        ) {
                            Text(stringResource(MR.strings.action_filter))
                        }
                    }

                    HorizontalDivider()
                }

                item {
                    SavedSearchItem(
                        savedSearches = savedSearches,
                        onSavedSearch = onSavedSearch,
                        onSavedSearchPress = onSavedSearchPress,
                        // KMK -->
                        onSavedSearchPressDesc = onSavedSearchPressDesc,
                        // KMK <--
                    )
                }

                items(filters) { filter ->
                    FilterItem(
                        filter = filter,
                        onUpdate = updateFilters,
                        // SY -->
                        startExpanded = startExpanded,
                        // SY <--
                    )
                }
            }
        },
    )
}

@Composable
private fun FilterItem(
    filter: Filter<*>,
    onUpdate: () -> Unit,
    // SY -->
    startExpanded: Boolean,
    // SY <--
) {
    when (filter) {
        // SY -->
        is Filter.AutoComplete -> {
            AutoCompleteItem(
                name = filter.name,
                state = filter.state.toImmutableList(),
                hint = filter.hint,
                values = filter.values.toImmutableList(),
                skipAutoFillTags = filter.skipAutoFillTags.toImmutableList(),
                validPrefixes = filter.validPrefixes.toImmutableList(),
                onChange = {
                    filter.state = it
                    onUpdate()
                },
            )
        }
        // SY <--

        is Filter.Header -> {
            HeadingItem(filter.name)
        }

        is Filter.Separator -> {
            HorizontalDivider()
        }

        is Filter.CheckBox -> {
            CheckboxItem(
                label = filter.name,
                checked = filter.state,
                onClick = {
                    filter.state = !filter.state
                    onUpdate()
                },
            )
        }

        is Filter.TriState -> {
            TriStateItem(
                label = filter.name,
                state = filter.state.toTriStateFilter(),
                onClick = { newState ->
                    filter.state = newState.toTriStateInt()
                    onUpdate()
                },
            )
        }

        is Filter.Text -> {
            TextItem(
                label = filter.name,
                value = filter.state,
                onChange = {
                    filter.state = it
                    onUpdate()
                },
            )
        }

        is Filter.Select<*> -> {
            SelectItem(
                label = filter.name,
                options = filter.values,
                selectedIndex = filter.state,
                onSelect = {
                    filter.state = it
                    onUpdate()
                },
            )
        }

        is Filter.Sort -> {
            CollapsibleBox(
                heading = filter.name,
                // SY -->
                startExpanded = startExpanded,
                // SY <--
                content = {
                    Column {
                        filter.values.forEachIndexed { index, item ->
                            val sortAscending = filter.state?.ascending
                                ?.takeIf { index == filter.state?.index }

                            SortItem(
                                label = item,
                                sortDescending = if (sortAscending != null) {
                                    !sortAscending
                                } else {
                                    null
                                },
                                onClick = {
                                    val ascending = if (index == filter.state?.index) {
                                        !filter.state!!.ascending
                                    } else {
                                        filter.state?.ascending ?: true
                                    }

                                    filter.state = Filter.Sort.Selection(
                                        index = index,
                                        ascending = ascending,
                                    )

                                    onUpdate()
                                },
                            )
                        }
                    }
                },
            )
        }

        is Filter.Group<*> -> {
            CollapsibleBox(
                heading = filter.name,
                // SY -->
                startExpanded = startExpanded,
                // SY <--
                content = {
                    Column {
                        filter.state
                            .filterIsInstance<Filter<*>>()
                            .forEach { childFilter ->
                                FilterItem(
                                    filter = childFilter,
                                    onUpdate = onUpdate,
                                    // SY -->
                                    startExpanded = startExpanded,
                                    // SY <--
                                )
                            }
                    }
                },
            )
        }
    }
}

private fun Int.toTriStateFilter(): TriState {
    return when (this) {
        Filter.TriState.STATE_IGNORE -> TriState.DISABLED
        Filter.TriState.STATE_INCLUDE -> TriState.ENABLED_IS
        Filter.TriState.STATE_EXCLUDE -> TriState.ENABLED_NOT
        else -> throw IllegalStateException("Unknown TriState state: $this")
    }
}

private fun TriState.toTriStateInt(): Int {
    return when (this) {
        TriState.DISABLED -> Filter.TriState.STATE_IGNORE
        TriState.ENABLED_IS -> Filter.TriState.STATE_INCLUDE
        TriState.ENABLED_NOT -> Filter.TriState.STATE_EXCLUDE
    }
}
